grails-app/domain/org/pih/warehouse/product/Product.groovy

```diff
@@ -616,17 +616,29 @@ class Product implements Comparable, Serializable {
      *
      * @return
      */
-    String toString() { return "${name}" }
+    @Override
+    String toString() { return name }
 
     /**
      * Sort by name
      */
+    @Override
     int compareTo(obj) {
+        return name <=> obj?.name ?: id <=> obj?.id
+    }
 
-        def sortOrder =
-                name <=> obj?.name ?:
-                        id <=> obj?.id
-        return sortOrder
+    @Override
+    int hashCode() {
+        return id?.hashCode() ?: super.hashCode()
```

Comment:
Some APIs use a product object as a key, and integration tests were failing to find the same object when looking up results.

In Grails 3 I'd love to make hashCode and equals a proper Trait and assign it to all domains to prevent bugs of this type, let me know what you think @jmiranda

Reply:
OBGM-341